docs/features/compilation-from-portable-pdb.md

```diff
@@ -0,0 +1,215 @@
+
+We will also store the MVID of a reference since it's a GUID that represents the symbol. This is to future proof the information we will need to look up references.
+
+At this time, we'll only include external references for the compilation. Any other references may be added later in a separate blob.
+
+## PDB Format Additions
+
+#### Metadata References
+
+Metadata references can be easily stored as binary. The binary encoding will be as follows (order matters):
+
+Name: A UTF-8 string (null terminated)
+
+Aliases: UTF-8 Comma (, ) separate list of aliases (null terminated). May be empty
+
+MetadataImageKind: byte value representing Microsoft. CodeAnalysis. MetadataImageKind
+
+EmbedInteropTypes/MetadataImageKind: byte, reads as binary values starting in order from right to left
+
+    
+    MetadataImageKind: 1 if Assembly, 0 if Module
+    EmbedInteropTypes: 1 if true
+    Examples:
+        0b11, MetadataImageKind.Assembly and EmbedInteropTypes = true
+        0b01, MetadataImageKind.Module and EmbedInteropTypes = true
+
+Timestamp: 4 byte integer
+
+File Size: 4 byte integer
+
+MVID: 16 byte integer (GUID)
+
+#### Compiler Flag Key Value Pairs
+
+The remaining values will be stored as key value pairs in the pdb. The storage format will be UTF8 encoded key value pairs that are null terminated. Order is not guaranteed. Any values left out can be assumed to be the default for the type. Keys may be different for Visual Basic and CSharp. They are serialized to reflect the command line arguments representing the same values
+
+Example: 
+
+`compilerversion\01.2.3\0sourceencoding\0utf-8\0checked\01\0unsafe\01\0langversion\0latest\0nullable\0Enable` 
+
+##### Options For CSharp
+
+See [compiler options](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/listed-alphabetically) documentation
+
+1. "compilerversion"
+2. "[codepage](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/codepage-compiler-option)"
+3. "[checked](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/checked-compiler-option)"
+4. "[unsafe](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/unsafe-compiler-option)"
+5. "[langversion](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/langversion-compiler-option)"
+6. "[nullable](https://docs.microsoft.com/en-us/dotnet/csharp/nullable-references)"
+7. "[define](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/define-compiler-option)"
+
+##### Options For Visual Basic
+
+See [compiler options](https://docs.microsoft.com/en-us/dotnet/visual-basic/reference/command-line-compiler/compiler-options-listed-alphabetically) documentation
+
+1. "compilerversion"
```

Comment:
>compilerversion [](start = 4, length = 15)

We also need to add `runtime-version`, which would be `(AssemblyInformationalVersionAttribute)typeof(object).Assembly.GetCustomAttributes(typeof(AssemblyInformationalVersionAttribute), false).Single().InformationalVersion`

On desktop .NET the version is e.g. `4.8.4180.0`, on .NET Core the version is e.g. `3.1.2-servicing.20066.2+fd22fd600a2526a7c58536da0e80325eb710d7f1`.

Reply:
Why do we need the runtime version in the PDB? Yeah there are some really odd corner cases where this would be relevant but in general it should not be. Seems odd to include this but not for instance the underlying operating system (Linux vs. Windows) which is much more likely to impact the deterministic output of the compilation.

Reply:
The runtime version affects Unicode character interpretation and decimal arithmetics. What differences are there between Linux and Windows? We expects paths to be normalized thru path map.

Reply:
The only OS-level differences I am aware of are in deflate compression implementation, since CLR currently calls into the OS to implement compression. This affects PDB embedded source files and embedded PDBs. Ideally, we would include the version of the native compression library as well. For that we will need API from CLR to retrieve the version. Alternatively, we can compare the decompressed versions of these artifacts for equality. But then we would also need to ignore PDB checksum when comparing the binaries.

Reply:
The tool that does the validation can download the specific version of CLR (or a container with that version of the CLR) and run in it. The same with compression lib, but it's harder to do. We could include the full OS version, but getting exact versions of various Linux distros is gonna be problematic (I think). Limiting the problem to finding a specific version of compression library seems easier.

Reply:
> What differences are there between Linux and Windows? 

The different line endings impacts determinism because it changes the checksum of source files across Linux / Windows. Even if you force consistent line endings in your source code across OS via .gitattributes (most repos I've seen don't do this) you still have to deal with the generated sources that get included in your builds. Last time I looked at this I saw generators using OS specific line endings which meant they would be different across Linux / Windows.

Reply:
The plan for line endings to record git options in Source Link: https://github.com/dotnet/sourcelink/issues/613. This should allow us to clone the repo with exactly the same line breaks that the compiler saw on the build server.

Reply:
I guess one disadvantage of including the runtime version is that now PDB/DLL actually depends on the runtime version even in cases when they did not before (similar problem like the encoding). We could potentially make this more fine-grained if the CLR exposed version of Unicode it supports and version of decimal operation implementation (not sure if that's actually viable). @richlander

Reply:
Thinking about this more - the usual case is that the compiler comes with the runtime in an SDK (on .NET Core), right? It is likely then that whenever compiler version changes the runtime changes as well.

Reply:
> The plan for line endings to record git options in Source Link: dotnet/sourcelink#613. This should allow us to clone the repo with exactly the same line breaks that the compiler saw on the build server

How will this fix generated files though? Think about generated assembly info. That comes from an MSBuild task which eventually (through `CodeDomProvider` and `TextWriter`) uses `Environment.NewLine` in the `AssemblyInfo.cs` file. That in turn will get included in the PDB file and seemingly (at least that I can think through) will have different line endings on Linux / Windows. 

Did want to ask, but your message seems to confirm, whether or not actually building repos was a goal. Seems like it is though.

Reply:
The files that are not checked into the repo are going to be embedded in the PDB. When reconstructing the compilation we are not gonna run the code generator or any msbuild targets. We are gonna take the files from the PDB and add them directly to the Compilation.

Reply:
> whether or not actually building repos was a goal.

It's not a goal to run msbuild. We are gonna bypass all msbuild and just create Compilation and call Emit on it to get the binary.